File: src/api/library_watch_notification.ts
// Private Instructor Watch inbox contracts.

import type { PublishedQuestionId } from "../../generated/api/PublishedQuestionId";

export type LibraryWatchTargetKind = "question" | "questionPool";

/** One self-only in-app notification; it contains no watcher or actor identity. */
type LibraryWatchNotificationBase = {
  readonly targetKind: LibraryWatchTargetKind;
  readonly targetPublicId: PublishedQuestionId;
  readonly occurredAt: number;
};

export type LibraryWatchNotification =
  | (LibraryWatchNotificationBase & {
      readonly eventKind: "revision";
      readonly revisionNumber: number;
      readonly forkedPublicId: null;
      readonly activityId: null;
    })
  | (LibraryWatchNotificationBase & {
      readonly eventKind: "fork";
      readonly revisionNumber: number;
      readonly forkedPublicId: PublishedQuestionId;
      readonly activityId: null;
    })
  | (LibraryWatchNotificationBase & {
      readonly eventKind: "improvementThread";
      /** Creation Revision and exact retained thread. */
      readonly revisionNumber: number;
      readonly forkedPublicId: null;
      readonly activityId: string;
    })
  | (LibraryWatchNotificationBase & {
      readonly eventKind: "impactNotice";
      /** The affected Revision can be absent; activityId is the exact retained notice. */
      readonly revisionNumber: number | null;
      readonly forkedPublicId: null;
      readonly activityId: string;
    });

export type LibraryWatchEventKind = LibraryWatchNotification["eventKind"];

export interface LibraryWatchNotificationClient {
  readonly getLibraryWatchNotifications: (
    limit?: number,
  ) => Promise<ReadonlyArray<LibraryWatchNotification>>;
}

File: crates/learning-data-access/src/library_watch_notification.rs
//! Private in-app Watch inbox and bounded materialization.
//!
//! This is the sole delivery boundary for Question and Question Pool Watch
//! activity. It intentionally has no email provider, recipient directory, or
//! public aggregate surface.

use async_trait::async_trait;
use question_model::{PublishedQuestionId, Timestamp};
use uuid::Uuid;

use crate::{SessionTokenHash, StoreError};

/// The two stable Library Object kinds a Watch can follow.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LibraryWatchTargetKind {
    Question,
    QuestionPool,
}

/// Complete evidence for one of the four events that reaches a Watch inbox.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LibraryWatchActivity {
    Revision {
        revision_number: u64,
    },
    Fork {
        source_revision_number: u64,
        forked_public_id: PublishedQuestionId,
    },
    ImprovementThread {
        creation_revision_number: u64,
        thread_id: Uuid,
    },
    ImpactNotice {
        affected_revision_number: Option<u64>,
        impact_notice_id: Uuid,
    },
}

/// One self-only, immutable in-app notification.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LibraryWatchNotification {
    pub target_kind: LibraryWatchTargetKind,
    pub target_public_id: PublishedQuestionId,
    pub activity: LibraryWatchActivity,
    pub occurred_at: Timestamp,
}

/// Least-privilege worker capability: materialize a bounded Watch outbox batch.
#[async_trait]
pub trait LibraryWatchNotificationStore: Send + Sync {
    async fn materialize_library_watch_notifications(&self, limit: u16) -> Result<u32, StoreError>;
}

/// Browser-facing private inbox for the authenticated active Instructor only.
#[async_trait]
pub trait LibraryWatchInboxStore: Send + Sync {
    async fn library_watch_notifications(
        &self,
        session_token_hash: SessionTokenHash,
        limit: u16,
    ) -> Result<Vec<LibraryWatchNotification>, StoreError>;
}
